SSA/Projects/InstCombine/AliveHandwrittenLargeExamples.lean

```diff
@@ -50,4 +61,102 @@ axioms: [propext, Classical.choice, Quot.sound] -/
 
 end DivRemOfSelect
 
+namespace AndOrXor
+
+open LLVMTheory
+
+/-
+Name: AndOrXor:2515   ((X^C1) >> C2)^C3 = (X>>C2) ^ ((C1>>C2)^C3)
+%e1  = xor %x, C1
+%op0 = lshr %e1, C2
+%r   = xor %op0, C3
+  =>
+%o = lshr %x, C2 -- (X>>C2)
+%p = lshr(%C1,%C2)
+%q = xor %p, %C3 -- ((C1>>C2)^C3)
+%r = xor %o, %q
+-/
+
+open ComWrappers
+def AndOrXor2515_lhs (w : ℕ):
+  Com InstCombine.LLVM
+    [/- C1 -/ InstCombine.Ty.bitvec w,
+     /- C2 -/ InstCombine.Ty.bitvec w,
+     /- C3 -/ InstCombine.Ty.bitvec w,
+     /- %X -/ InstCombine.Ty.bitvec w] (InstCombine.Ty.bitvec w) :=
+  /- e1  = -/ Com.lete (xor w /-x-/ 0 /-C1-/ 3) <|
+  /- op0 = -/ Com.lete (lshr w /-e1-/ 0 /-C2-/ 3) <|
+  /- r   = -/ Com.lete (xor w /-op0-/ 0 /-C3-/ 3) <|
+  Com.ret ⟨/-r-/0, by simp [Ctxt.snoc]⟩
+
+def AndOrXor2515_rhs (w : ℕ) :
+  Com InstCombine.LLVM
+    [/- C1 -/ InstCombine.Ty.bitvec w,
+     /- C2 -/ InstCombine.Ty.bitvec w,
+     /- C3 -/ InstCombine.Ty.bitvec w,
+     /- %X -/ InstCombine.Ty.bitvec w] (InstCombine.Ty.bitvec w) :=
+  /- o = -/ Com.lete (lshr w /-X-/ 0 /-C2-/ 2) <|
+  /- p = -/ Com.lete (lshr w /-C1-/ 4 /-C2-/ 3) <|
+  /- q = -/ Com.lete (xor w /-p-/ 0 /-C3-/ 3) <|
+  /- r = -/ Com.lete (xor w /-o-/ 2 /-q-/ 0) <|
+  Com.ret ⟨/-r-/0, by simp [Ctxt.snoc]⟩
+
+def ushr_xor_right_distrib (c1 c2 c3 : BitVec w): (c1 ^^^ c2) >>> c3 = (c1 >>> c3) ^^^ (c2 >>> c3) := by
+  unfold HShiftRight.hShiftRight instHShiftRightBitVec
+  ext
+  simp [getLsb_ushiftRight]
+
+def xor_assoc (c1 c2 c3 : BitVec w): c1 ^^^ c2 ^^^ c3 = c1 ^^^ (c2 ^^^ c3) := by
+  ext i
+  simp
+
+def help'' {a b : Nat } : a = b ↔ (↑a : Int) = ↑b := by
+    simp only [Nat.cast_inj]
+
+def alive_simplifyAndOrXor2515 (w : Nat) :
+  AndOrXor2515_lhs w ⊑ AndOrXor2515_rhs w := by
+  simp only [AndOrXor2515_lhs, AndOrXor2515_rhs]
+  simp only [simp_llvm_wrap]
+  simp_alive_ssa
+  simp_alive_undef
+  intros c1 c2 c3 x
+  rcases c1 with none | c1 <;>
+  rcases c2 with none | c2 <;>
+  rcases c3 with none | c3 <;>
+  rcases x with none | x <;>
+  simp only [LLVM.xor?_eq, xor_eq, Option.bind_eq_bind, Option.none_bind, Option.bind_none,
+    Option.some_bind, Refinement.refl]
+  rw [←Option.bind_eq_bind]
+  simp_alive_ops
+  by_cases h : w ≤ BitVec.toNat c2 <;>
+  simp only [ge_iff_le, h, ↓reduceIte, Option.bind_eq_bind, Option.none_bind, Option.bind_none,
+    Refinement.refl, Option.some_bind, h, Option.pure_def, Option.some_bind, Refinement.some_some]
+  simp only [ushr_xor_right_distrib, xor_assoc]
```

Comment:
Could have sworn you had a cleaner proof @tobiasgrosser :)

Reply:
A managed to clean it up a bit, but not by much.